[13.0][IMP] shopfloor: checkout, use of delivery packages

sebalix opened the pull request:
<s>PR based on https://github.com/OCA/wms/pull/193, only the two last commits are relevant.</s>

A delivery package is a package with a delivery packaging defined ('packaging_id' field). It is a package different than internal ones (which are stored in the warehouse with their own constraints like height...).
Here we enforce the use of such delivery packages when we perform a checkout, this means that we can not use internal packages to perform a delivery.

* `list_dest_package` method ('/select_dest_package') is now listing only delivery packages
* `scan_dest_package` and 'set_dest_package' methods accept only a delivery package
* `scan_package_action` method accept only a delivery package
* Add a new state/screen when the user click on the "_New package_" button to list available delivery packaging (`select_delivery_packaging` state) for the current transfer and select one to create the package (existing `new_package` method which now accepts a packaging as parameter).

Ref. 2168

sebalix committed:
shopfloor: checkout, allows delivery packages only

A delivery package is a package with a delivery packaging defined
('packaging_id' field). It is a package different than internal ones
(which are stored in the warehouse with their own constraints like
height...).
Here we enforce the use of such delivery packages when we perform a
checkout, this means that we can not use internal packages to perform a
delivery.

* 'list_dest_package' method ('/select_dest_package') is now listing only
  delivery packages
* 'scan_dest_package' and 'set_dest_package' methods accept only a
  delivery package
* 'scan_package_action' method accept only a delivery package

sebalix committed:
shopfloor: checkout, select the delivery packaging

Add a new state/screen when the user click on the "New package" button
to list available delivery packaging ('select_delivery_packaging' state)
for the current transfer and select one to create the package (existing
'new_package' method which now accepts a packaging as parameter).

simahawk commented on shopfloor/services/checkout.py:
```diff
@@ -120,6 +122,19 @@ def _response_for_select_dest_package(self, picking, move_lines, message=None):
+            next_state="select_delivery_packaging",
+            data={
+                # We don't need to send the 'picking' as the mobile frontend
+                # already this info
```
```suggestion
                # already has this info after `select_document` state
```

simahawk commented on shopfloor/services/checkout.py:
```diff
@@ -619,26 +637,28 @@ def _filter_lines_checkout_done(move_line):
 
     def _is_package_allowed(self, picking, package):
-        existing_packages = picking.mapped("move_line_ids.result_package_id")
+        """A package is allowed as a destination one if it is present among
```
https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

simahawk commented on shopfloor/services/checkout.py:
```diff
@@ -678,10 +698,10 @@ def _create_and_assign_new_packaging(self, picking, selected_lines, packaging=No
+        When a package is scanned (only delivery ones), if the package is known
+        as the destination package of one of the lines or is the source package
+        of a selected line, the package is set to be the destination package of
+        all then lines to pack.
```
```suggestion
        all the lines to pack.
```

simahawk commented on shopfloor/services/checkout.py:
```diff
@@ -758,7 +784,42 @@ def scan_package_action(self, picking_id, selected_line_ids, barcode):
+        )
+
+    def list_delivery_packaging(self, picking_id, selected_line_ids):
+        """List available delivery packaging based on the carrier of the
+        picking to create the new package.
```
```suggestion
        """List available delivery packaging for given picking.

```

simahawk commented on shopfloor/services/checkout.py:
```diff
@@ -823,17 +887,20 @@ def list_dest_package(self, picking_id, selected_line_ids):
+            return self._response_for_select_dest_package(
+                picking,
+                selected_lines,
+                message=self.msg_store.package_not_found_for_barcode(barcode),
```
what if barcode is none?

simahawk replied:
probably you should not call `_set_dest_package_from_selection` if you don't have a package

sebalix replied:
I'll check this point.

sebalix replied:
Yes it's a bit hackish, but this happens only when `_set_dest_package_from_selection` is called by `set_dest_package` (which has a `package_id` as parameter), so the only way to have `barcode=None` here is we send a wrong `package_id` to `set_dest_package`, and then yes the error message is not really appropriate (I talked about this in the related test).
Maybe the best way here is to return the generic "record not found" error message before calling `_set_dest_package_from_selection` yes.

sebalix replied:
Fixed, we now test that the package exists in `set_dest_package` and `scan_dest_package` before calling `_set_dest_package_from_selection`. The error message differs between the two methods if the package can't be found.

simahawk commented on shopfloor/services/checkout.py:
```diff
@@ -1277,6 +1356,15 @@ def _schema_select_package(self):
+    @property
+    def _schema_select_delivery_packaging(self):
+        return {
+            "picking": {"type": "dict", "schema": self.schemas.picking()},
```
```suggestion
            "picking": self.schemas._schema_dict_of(self.schemas.picking()),
```

simahawk commented on shopfloor/tests/test_checkout_list_delivery_packaging.py:
```diff
@@ -0,0 +1,122 @@
+# Copyright 2020 Camptocamp SA (http://www.camptocamp.com)
```
2021

simahawk commented on shopfloor/tests/test_checkout_list_package.py:
```diff
@@ -115,29 +117,42 @@ def setUpClassBaseData(cls):
         cls.picking = picking
-        cls.package = cls.move_line1.result_package_id
+
+    def _get_allowed_packages(self):
```
pass the picking instead of using self.picking

simahawk commented:
#193 merged

simahawk commented on shopfloor_mobile/static/wms/src/scenario/checkout.js:
```diff
@@ -257,6 +271,14 @@ const Checkout = {
+            return {
+                list_item_options: {
+                    loud_title: true,
+                    fields: [{path: "packaging_type"}, {path: "shipper_package_code"}],
```
for `shipper_package_code` maybe better to show a label?

sebalix replied:
done, same label than the backend

simahawk reviewed (commented):
my PR has been merged, rebase pls :)

sebalix commented:
@simahawk don't merge now, following my test with @jbaudoux I have to change small things regarding this (and some other fixes not related to this PR particularly)

sebalix committed:
shopfloor: checkout, select line after canceling

Once a line or package has been canceled, return on the `select_line`
state instead of `summary`.
When we cancel a line or a package, there is a high chance that the user
wants to restart again the operation.

sebalix commented:
Changes applied since the last push:

- list of delivery packaging are now sorted by name
- on the list of delivery packaging, no need to click on the _"Confirm"_ button once the packaging has been selected: it is chosen directly when selected
- on the list of delivery packaging, the carrier is displayed like the other screens to help the user to choose the right packaging if needed
- remove of `shipper_package_code` data on listed delivery packaging (not useful)
- change the error message _"Packaging {} does not match carrier {}."_ en _"Packaging {} is not allowed for carrier {}."_ when we scan a delivery packaging not related to the current carrier
- when we cancel a line or a package, get back on the `select_line` screen instead of the `summary` one

Ready for reviews/merge.

simahawk commented:
/ocabot merge minor
